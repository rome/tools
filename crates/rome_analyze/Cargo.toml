[package]
name = "rome_analyze"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rome_rowan = { path = "../rome_rowan" }
rome_control_flow = { path = "../rome_control_flow" }
rome_console = { path = "../rome_console" }
rome_diagnostics = { path = "../rome_diagnostics" }
bitflags = "1.3.2"

[dev-dependencies]
rome_js_syntax = { path = "../rome_js_syntax" }
rome_js_factory = { path = "../rome_js_factory" }
