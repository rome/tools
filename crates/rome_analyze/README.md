# Rome Analyze

## This entire crate is a WIP with a clunky and unstable API. 

The data structures and methods are poorly designed/named and will certainly change.

- `analyzers` produce diagnostics and fixes for an entire file and are relevant both for LSP and CLI use.
- `assists` produce code actions related to cursor position and are only relevant for LSP.