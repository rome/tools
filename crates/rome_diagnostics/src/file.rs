use rome_console::codespan::SourceFile;
use rome_rowan::{Language, SyntaxElement, SyntaxNode, SyntaxToken, TextRange, TextSize};
use std::{collections::HashMap, fmt::Debug, ops::Range};

/// A value which can be used as the range inside of a diagnostic.
///
/// This is essentially a hack to allow us to use SyntaxElement, SyntaxNode, etc directly
pub trait Span {
    fn as_range(&self) -> TextRange;

    /// Make a new span which extends to another span
    ///
    /// ```text
    /// from      to
    /// ^^^^^^^^^^^^
    /// ```
    fn join<T: Span>(&self, other: T) -> TextRange {
        TextRange::new(self.as_range().start(), other.as_range().end())
    }

    /// Make a new span which is between another span
    ///
    /// ```text
    /// from      to
    ///     ^^^^^^
    /// ```
    fn between<T: Span>(&self, other: T) -> TextRange {
        TextRange::new(self.as_range().end(), other.as_range().start())
    }

    /// Make a new span which extends until another span
    ///
    /// ```text
    /// from      to
    /// ^^^^^^^^^^
    /// ```
    fn until<T: Span>(&self, other: T) -> TextRange {
        TextRange::new(self.as_range().start(), other.as_range().start())
    }

    fn sub_start(&self, amount: TextSize) -> TextRange {
        let range = self.as_range();
        TextRange::new(range.start() - amount, range.end())
    }

    fn add_start(&self, amount: TextSize) -> TextRange {
        let range = self.as_range();
        TextRange::new(range.start() + amount, range.end())
    }

    fn sub_end(&self, amount: TextSize) -> TextRange {
        let range = self.as_range();
        TextRange::new(range.start(), range.end() - amount)
    }

    fn add_end(&self, amount: TextSize) -> TextRange {
        let range = self.as_range();
        TextRange::new(range.start(), range.end() + amount)
    }
}

impl<T: Span> Span for &T {
    fn as_range(&self) -> TextRange {
        (*self).as_range()
    }
}

impl<T: Span> Span for &mut T {
    fn as_range(&self) -> TextRange {
        (**self).as_range()
    }
}

impl<T: Copy> Span for Range<T>
where
    TextSize: TryFrom<T>,
    <TextSize as TryFrom<T>>::Error: Debug,
{
    fn as_range(&self) -> TextRange {
        TextRange::new(
            TextSize::try_from(self.start).expect("integer overflow"),
            TextSize::try_from(self.end).expect("integer overflow"),
        )
    }
}

impl<T: Language> Span for SyntaxNode<T> {
    fn as_range(&self) -> TextRange {
        self.text_range()
    }
}

impl<T: Language> Span for SyntaxToken<T> {
    fn as_range(&self) -> TextRange {
        self.text_range()
    }
}

impl<T: Language> Span for SyntaxElement<T> {
    fn as_range(&self) -> TextRange {
        match self {
            SyntaxElement::Node(n) => n.text_range(),
            SyntaxElement::Token(t) => t.text_range(),
        }
    }
}

impl Span for TextRange {
    fn as_range(&self) -> TextRange {
        *self
    }
}

/// An id that points into a file database.
pub type FileId = usize;

/// A range that is indexed in a specific file.
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct FileSpan {
    pub file: FileId,
    pub range: TextRange,
}

impl FileSpan {
    pub fn new(file: FileId, span: impl Span) -> Self {
        let range = span.as_range();
        Self { file, range }
    }
}

/// Interface for interacting with source files
/// that are identified by a unique identifier.
pub trait Files {
    /// Returns the name of the file identified by the id.
    fn name(&self, id: FileId) -> Option<&str>;

    /// Returns the source of the file identified by the id.
    fn source(&self, id: FileId) -> Option<SourceFile<'_>>;
}

/// A file database that contains only one file.
#[derive(Clone, Debug)]
pub struct SimpleFile {
    name: String,
    source: String,
    line_starts: Vec<TextSize>,
}

impl SimpleFile {
    /// Create a new file with the name and source.
    pub fn new(name: String, source: String) -> Self {
        Self {
            line_starts: SourceFile::line_starts(&source).collect(),
            name,
            source,
        }
    }

    /// Returns a `SimpleFile` that has no name and no source.
    pub fn empty() -> SimpleFile {
        SimpleFile::new(String::new(), String::new())
    }
}

impl Files for SimpleFile {
    fn name(&self, _id: FileId) -> Option<&str> {
        Some(&self.name)
    }

    fn source(&self, _id: FileId) -> Option<SourceFile<'_>> {
        Some(SourceFile::new(&self.source, &self.line_starts))
    }
}

/// A file database that stores multiple files.
#[derive(Clone, Debug, Default)]
pub struct SimpleFiles {
    files: HashMap<FileId, SimpleFile>,
    id: usize,
}

impl SimpleFiles {
    pub fn new() -> Self {
        Self::default()
    }

    /// Adds a file to this database and returns the id for the new file.
    pub fn add(&mut self, name: String, source: String) -> FileId {
        let id = self.id;
        self.id += 1;
        self.files.insert(id, SimpleFile::new(name, source));
        id
    }

    pub fn get(&self, id: FileId) -> Option<&SimpleFile> {
        self.files.get(&id)
    }
}

impl Files for SimpleFiles {
    fn name(&self, id: FileId) -> Option<&str> {
        self.files.get(&id)?.name(id)
    }

    fn source(&self, id: FileId) -> Option<SourceFile<'_>> {
        self.files.get(&id)?.source(id)
    }
}
