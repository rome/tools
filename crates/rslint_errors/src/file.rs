use rome_rowan::{Language, SyntaxElement, SyntaxNode, SyntaxToken, TextRange};
use std::{collections::HashMap, ops::Range};

/// A value which can be used as the range inside of a diagnostic.
///
/// This is essentially a hack to allow us to use SyntaxElement, SyntaxNode, etc directly
pub trait Span {
    fn as_range(&self) -> Range<usize>;

    fn as_text_range(&self) -> TextRange {
        TextRange::new(
            (self.as_range().start as u32).into(),
            (self.as_range().end as u32).into(),
        )
    }

    /// Make a new span which extends to another span
    ///
    /// ```text
    /// from      to
    /// ^^^^^^^^^^^^
    /// ```
    fn join<T: Span>(&self, other: T) -> Range<usize> {
        self.as_range().start..other.as_range().end
    }

    /// Make a new span which is between another span
    ///
    /// ```text
    /// from      to
    ///     ^^^^^^
    /// ```
    fn between<T: Span>(&self, other: T) -> Range<usize> {
        self.as_range().end..other.as_range().start
    }

    /// Make a new span which extends until another span
    ///
    /// ```text
    /// from      to
    /// ^^^^^^^^^^
    /// ```
    fn until<T: Span>(&self, other: T) -> Range<usize> {
        self.as_range().start..other.as_range().start
    }

    fn sub_start(&self, amount: usize) -> Range<usize> {
        let range = self.as_range();
        range.start - amount..range.end
    }

    fn add_start(&self, amount: usize) -> Range<usize> {
        let range = self.as_range();
        range.start + amount..range.end
    }

    fn sub_end(&self, amount: usize) -> Range<usize> {
        let range = self.as_range();
        range.start..range.end - amount
    }

    fn add_end(&self, amount: usize) -> Range<usize> {
        let range = self.as_range();
        range.start..range.end + amount
    }
}

impl<T: Span> Span for &T {
    fn as_range(&self) -> Range<usize> {
        (*self).as_range()
    }
}

impl<T: Span> Span for &mut T {
    fn as_range(&self) -> Range<usize> {
        (**self).as_range()
    }
}

impl<T: Clone> Span for Range<T>
where
    T: Into<usize>,
{
    fn as_range(&self) -> Range<usize> {
        self.start.clone().into()..self.end.clone().into()
    }
}

impl<T: Language> Span for SyntaxNode<T> {
    fn as_range(&self) -> Range<usize> {
        self.text_range().into()
    }
}

impl<T: Language> Span for SyntaxToken<T> {
    fn as_range(&self) -> Range<usize> {
        self.text_range().into()
    }
}

impl<T: Language> Span for SyntaxElement<T> {
    fn as_range(&self) -> Range<usize> {
        match self {
            SyntaxElement::Node(n) => n.text_range(),
            SyntaxElement::Token(t) => t.text_range(),
        }
        .into()
    }
}

impl Span for TextRange {
    fn as_range(&self) -> Range<usize> {
        (*self).into()
    }
}

/// An id that points into a file database.
pub type FileId = usize;

/// A range that is indexed in a specific file.
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct FileSpan {
    pub file: FileId,
    pub range: Range<usize>,
}

impl FileSpan {
    pub fn new(file: FileId, span: impl Span) -> Self {
        let range = span.as_range();
        debug_assert!(
            range.start <= range.end,
            "slice index starts at {} but ends at {}",
            range.start,
            range.end
        );

        Self { file, range }
    }
}

/// Interface for interacting with source files
/// that are identified by a unique identifier.
pub trait Files {
    /// Returns the name of the file identified by the id.
    fn name(&self, id: FileId) -> Option<&str>;

    /// Returns the source of the file identified by the id.
    fn source(&self, id: FileId) -> Option<&str>;

    /// The index of the line at the byte index.
    ///
    /// ## Implementation
    /// This can be implemented by caching the results of [`line_starts`]
    /// and then use [`binary_search`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search)
    /// to compute the line index.
    ///
    /// ```ignore
    /// match self.line_starts.binary_search(byte_index) {
    ///     Ok(line) => line,
    ///     Err(next_line) => next_line - 1,
    /// }
    /// ```
    fn line_index(&self, file_id: FileId, byte_index: usize) -> Option<usize>;

    /// The byte range of line in the source of the file.
    fn line_range(&self, id: FileId, line_index: usize) -> Option<Range<usize>>;
}

/// A file database that contains only one file.
#[derive(Clone, Debug)]
pub struct SimpleFile {
    name: String,
    source: String,
    line_starts: Vec<usize>,
}

impl SimpleFile {
    /// Create a new file with the name and source.
    pub fn new(name: String, source: String) -> Self {
        Self {
            line_starts: line_starts(&source).collect(),
            name,
            source,
        }
    }

    /// Returns a `SimpleFile` that has no name and no source.
    pub fn empty() -> SimpleFile {
        SimpleFile::new(String::new(), String::new())
    }

    fn line_start(&self, line_index: usize) -> Option<usize> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => self.line_starts.get(line_index).cloned(),
            Ordering::Equal => Some(self.source.len()),
            Ordering::Greater => None,
        }
    }
}

impl Files for SimpleFile {
    fn name(&self, _id: FileId) -> Option<&str> {
        Some(&self.name)
    }

    fn source(&self, _id: FileId) -> Option<&str> {
        Some(&self.source)
    }

    fn line_index(&self, _file_id: FileId, byte_index: usize) -> Option<usize> {
        Some(
            self.line_starts
                .binary_search(&byte_index)
                .unwrap_or_else(|next_line| next_line - 1),
        )
    }

    fn line_range(&self, _: FileId, line_index: usize) -> Option<Range<usize>> {
        let line_start = self.line_start(line_index)?;
        let next_line_start = self.line_start(line_index + 1)?;

        Some(line_start..next_line_start)
    }
}

/// A file database that stores multiple files.
#[derive(Clone, Debug, Default)]
pub struct SimpleFiles {
    files: HashMap<FileId, SimpleFile>,
    id: usize,
}

impl SimpleFiles {
    pub fn new() -> Self {
        Self::default()
    }

    /// Adds a file to this database and returns the id for the new file.
    pub fn add(&mut self, name: String, source: String) -> FileId {
        let id = self.id;
        self.id += 1;
        self.files.insert(id, SimpleFile::new(name, source));
        id
    }

    pub fn get(&self, id: FileId) -> Option<&SimpleFile> {
        self.files.get(&id)
    }
}

impl Files for SimpleFiles {
    fn name(&self, id: FileId) -> Option<&str> {
        self.files.get(&id)?.name(id)
    }

    fn source(&self, id: FileId) -> Option<&str> {
        self.files.get(&id)?.source(id)
    }

    fn line_index(&self, id: FileId, byte_index: usize) -> Option<usize> {
        self.files.get(&id)?.line_index(id, byte_index)
    }

    fn line_range(&self, file_id: FileId, line_index: usize) -> Option<Range<usize>> {
        self.files.get(&file_id)?.line_range(file_id, line_index)
    }
}

/// Computes the byte indicies of every line start.
pub fn line_starts(source: &str) -> impl '_ + Iterator<Item = usize> {
    std::iter::once(0).chain(
        source
            .match_indices(&['\n', '\r', '\u{2028}', '\u{2029}'][..])
            .map(|(i, _)| i + 1),
    )
}
