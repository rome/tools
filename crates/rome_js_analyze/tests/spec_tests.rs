use std::{
    ffi::OsStr, fmt::Write, fs::read_to_string, os::raw::c_int, path::Path, slice, sync::Once,
};

use similar::TextDiff;

use rome_analyze::{
    AnalysisFilter, AnalyzerAction, AnalyzerOptions, ControlFlow, Never, RuleFilter,
};
use rome_console::{
    fmt::{Formatter, Termcolor},
    markup, Markup,
};
use rome_diagnostics::file::FileId;
use rome_diagnostics::Severity;
use rome_diagnostics::{file::SimpleFile, termcolor::NoColor, Diagnostic};
use rome_js_parser::{
    parse,
    test_utils::{assert_errors_are_absent, has_unknown_nodes_or_empty_slots},
};
use rome_js_syntax::{JsLanguage, SourceType};

tests_macros::gen_tests! {"tests/specs/**/*.{cjs,js,jsx,tsx,ts}", crate::run_test, "module"}

fn run_test(input: &'static str, _: &str, _: &str, _: &str) {
    register_leak_checker();

    let input_file = Path::new(input);
    let file_name = input_file.file_name().and_then(OsStr::to_str).unwrap();
    let input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {:?}: {:?}", input_file, err));

    let source_type = input_file.try_into().unwrap();
    let parsed = parse(&input_code, FileId::zero(), source_type);
    let root = parsed.tree();

    let (group, rule) = parse_test_path(input_file);

    let rule_filter = RuleFilter::Rule(group, rule);
    let filter = AnalysisFilter {
        enabled_rules: Some(slice::from_ref(&rule_filter)),
        ..AnalysisFilter::default()
    };

    let mut diagnostics = Vec::new();
    let mut code_fixes = Vec::new();
    let options = AnalyzerOptions::default();
    rome_js_analyze::analyze(FileId::zero(), &root, filter, &options, |event| {
        if let Some(diag) = event.diagnostic() {
            let mut diag = diag.into_diagnostic(Severity::Warning);
            if let Some(action) = event.action() {
                check_code_action(input_file, &input_code, source_type, &action);
                diag.suggestions.push(action.into());
            }

            diagnostics.push(diagnostic_to_string(file_name, &input_code, diag));
            return ControlFlow::Continue(());
        }

        if let Some(action) = event.action() {
            check_code_action(input_file, &input_code, source_type, &action);
            code_fixes.push(code_fix_to_string(&input_code, action));
        }

        ControlFlow::<Never>::Continue(())
    });

    let mut snapshot = String::new();

    writeln!(snapshot, "# Input").unwrap();
    writeln!(snapshot, "```js").unwrap();
    writeln!(snapshot, "{}", input_code).unwrap();
    writeln!(snapshot, "```").unwrap();
    writeln!(snapshot).unwrap();

    if !diagnostics.is_empty() {
        writeln!(snapshot, "# Diagnostics").unwrap();
        for diagnostic in diagnostics {
            writeln!(snapshot, "```").unwrap();
            writeln!(snapshot, "{}", diagnostic).unwrap();
            writeln!(snapshot, "```").unwrap();
            writeln!(snapshot).unwrap();
        }
    }

    if !code_fixes.is_empty() {
        writeln!(snapshot, "# Actions").unwrap();
        for action in code_fixes {
            writeln!(snapshot, "```diff").unwrap();
            writeln!(snapshot, "{}", action).unwrap();
            writeln!(snapshot, "```").unwrap();
            writeln!(snapshot).unwrap();
        }
    }

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(file_name, snapshot, file_name);
    });
}

/// The test runner for the analyzer is currently designed to have a
/// one-to-one mapping between test case and analyzer rules, so each testing
/// file will be run through the analyzer with only the rule corresponding
/// to the file name (or the name of the parent directory if it's not "specs")
/// enabled, eg. `correctness/useWhile.js` and `correctness/useWhile/test.js` will be analyzed with
/// just the `correctness-ignore lint(correctness/useW/useWhile` rule
fn parse_test_path(file: &Path) -> (&str, &str) {
    let file_stem = file.file_stem().unwrap();

    let ancestor_0 = file.parent().unwrap();
    let name_0 = ancestor_0.file_name().unwrap();

    let ancestor_1 = ancestor_0.parent().unwrap();
    let name_1 = ancestor_1.file_name().unwrap();

    if name_1.to_str().unwrap() == "specs" {
        (name_0.to_str().unwrap(), file_stem.to_str().unwrap())
    } else {
        (name_1.to_str().unwrap(), name_0.to_str().unwrap())
    }
}

fn markup_to_string(markup: Markup) -> String {
    let mut buffer = Vec::new();
    let mut write = Termcolor(NoColor::new(&mut buffer));
    let mut fmt = Formatter::new(&mut write);
    fmt.write_markup(markup).unwrap();

    String::from_utf8(buffer).unwrap()
}
#[allow(clippy::let_and_return)]
fn diagnostic_to_string(name: &str, source: &str, diag: Diagnostic) -> String {
    let file = SimpleFile::new(name.into(), source.into());
    let text = markup_to_string(markup! {
        {diag.display(&file)}
    });

    text
}

fn check_code_action(
    path: &Path,
    source: &str,
    source_type: SourceType,
    action: &AnalyzerAction<JsLanguage>,
) {
    let (_, text_edit) = action.mutation.as_text_edits().unwrap_or_default();

    let output = text_edit.new_string(source);

    let new_tree = action.mutation.clone().commit();

    // Checks that applying the text edits returned by the BatchMutation
    // returns the same code as printing the modified syntax tree
    assert_eq!(new_tree.to_string(), output);

    if has_unknown_nodes_or_empty_slots(&new_tree) {
        panic!("modified tree has unknown nodes or empty slots:\n{new_tree:#?}")
    }

    // Checks the returned tree contains no missing children node
    if format!("{new_tree:?}").contains("missing (required)") {
        panic!("modified tree has missing children:\n{new_tree:#?}")
    }

    // Re-parse the modified code and panic if the resulting tree has syntax errors
    let re_parse = parse(&output, FileId::zero(), source_type);
    assert_errors_are_absent(&re_parse, path);
}

fn code_fix_to_string(source: &str, action: AnalyzerAction<JsLanguage>) -> String {
    let (_, text_edit) = action.mutation.as_text_edits().unwrap_or_default();

    let output = text_edit.new_string(source);

    let diff = TextDiff::from_lines(source, &output);

    let mut diff = diff.unified_diff();
    diff.context_radius(3);

    diff.to_string()
}

// Check that all red / green nodes have correctly been released on exit
extern "C" fn check_leaks() {
    if let Some(report) = rome_rowan::check_live() {
        panic!("\n{report}")
    }
}

fn register_leak_checker() {
    // Import the atexit function from libc
    extern "C" {
        fn atexit(f: extern "C" fn()) -> c_int;
    }

    // Use an atomic Once to register the check_leaks function to be called
    // when the process exits
    static ONCE: Once = Once::new();
    ONCE.call_once(|| unsafe {
        countme::enable(true);
        atexit(check_leaks);
    });
}
