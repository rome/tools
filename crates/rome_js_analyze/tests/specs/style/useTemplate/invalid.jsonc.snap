---
source: crates/rome_js_analyze/tests/spec_tests.rs
expression: invalid.jsonc
---
# Input
```js
const foo = 'bar';
```

# Input
```js
console.log(foo + 'baz');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log(foo + 'baz');
      │             ^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log(foo·+·'baz');
  + console.log(`${foo}baz`);
  

```

# Input
```js
console.log(1 * 2 + 'foo');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log(1 * 2 + 'foo');
      │             ^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log(1·*·2·+·'foo');
  + console.log(`${1·*·2}foo`);
  

```

# Input
```js
console.log(1 + 'foo' + 2 + 'bar' + 'baz' + 3);
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log(1 + 'foo' + 2 + 'bar' + 'baz' + 3);
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log(1·+·'foo'·+·2·+·'bar'·+·'baz'·+·3);
  + console.log(`${1}foo${2}barbaz${3}`);
  

```

# Input
```js
console.log((1 + 'foo') * 2);
```

# Diagnostics
```
invalid.jsonc:1:14 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log((1 + 'foo') * 2);
      │              ^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log((1·+·'foo')·*·2);
  + console.log((`${1}foo`)·*·2);
  

```

# Input
```js
console.log(1 * (2 + 'foo') + 'bar');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log(1 * (2 + 'foo') + 'bar');
      │             ^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log(1·*·(2·+·'foo')·+·'bar');
  + console.log(`${1·*·(2·+·'foo')}bar`);
  

```

```
invalid.jsonc:1:18 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log(1 * (2 + 'foo') + 'bar');
      │                  ^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log(1·*·(2·+·'foo')·+·'bar');
  + console.log(1·*·(`${2}foo`)·+·'bar');
  

```

# Input
```js
console.log('foo' + 1);
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo' + 1);
      │             ^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo'·+·1);
  + console.log(`foo${1}`);
  

```

# Input
```js
console.log('foo' + `bar${`baz${'bat' + 'bam'}`}` + 'boo');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo' + `bar${`baz${'bat' + 'bam'}`}` + 'boo');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo'·+·`bar${`baz${'bat'·+·'bam'}`}`·+·'boo');
  + console.log(`foobarbaz${'bat'·+·'bam'}boo`);
  

```

# Input
```js
console.log('foo' + 1 + 2);
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo' + 1 + 2);
      │             ^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo'·+·1·+·2);
  + console.log(`foo${1}${2}`);
  

```

# Input
```js
1 + '2' - 3;
```

# Diagnostics
```
invalid.jsonc:1:1 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ 1 + '2' - 3;
      │ ^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - 1·+·'2'·-·3;
  + `${1}2`·-·3;
  

```

# Input
```js
foo() + ' bar';
```

# Diagnostics
```
invalid.jsonc:1:1 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ foo() + ' bar';
      │ ^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - foo()·+·'·bar';
  + `${foo()}·bar`;
  

```

# Input
```js
1 * /**leading*/'foo'    /**trailing */                   + 'bar'
```

# Diagnostics
```
invalid.jsonc:1:1 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ 1 * /**leading*/'foo'    /**trailing */                   + 'bar'
      │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - 1·*·/**leading*/'foo'····/**trailing·*/···················+·'bar'
  + `${1·*·/**leading*/'foo'····/**trailing·*/}bar`
  

```

# Input
```js
console.log('${foo.' + bar + '.baz}');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('${foo.' + bar + '.baz}');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('${foo.'·+·bar·+·'.baz}');
  + console.log(`\${foo.${bar}.baz}`);
  

```

# Input
```js
console.log('foo: ${bar.' + baz + '.bat}');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo: ${bar.' + baz + '.bat}');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo:·${bar.'·+·baz·+·'.bat}');
  + console.log(`foo:·\${bar.${baz}.bat}`);
  

```

# Input
```js
console.log('foo: `bar.' + baz + '.bat}');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo: `bar.' + baz + '.bat}');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo:·`bar.'·+·baz·+·'.bat}');
  + console.log(`foo:·\`bar.${baz}.bat}`);
  

```

# Input
```js
console.log('${foo}: `bar.' + baz + '.bat}');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('${foo}: `bar.' + baz + '.bat}');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('${foo}:·`bar.'·+·baz·+·'.bat}');
  + console.log(`\${foo}:·\`bar.${baz}.bat}`);
  

```

# Input
```js
console.log('foo: ${bar.' + baz + '.bat}');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo: ${bar.' + baz + '.bat}');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo:·${bar.'·+·baz·+·'.bat}');
  + console.log(`foo:·\${bar.${baz}.bat}`);
  

```

# Input
```js
console.log('foo: `bar.' + baz + '.bat}');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo: `bar.' + baz + '.bat}');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo:·`bar.'·+·baz·+·'.bat}');
  + console.log(`foo:·\`bar.${baz}.bat}`);
  

```

# Input
```js
console.log('foo: \${bar.' + baz + '.bat}');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo: \${bar.' + baz + '.bat}');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo:·\${bar.'·+·baz·+·'.bat}');
  + console.log(`foo:·\${bar.${baz}.bat}`);
  

```

# Input
```js
console.log('foo: \${bar.' + baz + '.bat}');
```

# Diagnostics
```
invalid.jsonc:1:13 lint/style/useTemplate  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Template literals are preferred over string concatenation.
  
  > 1 │ console.log('foo: \${bar.' + baz + '.bat}');
      │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Suggested fix: Use a TemplateLiteral.
  
  - console.log('foo:·\${bar.'·+·baz·+·'.bat}');
  + console.log(`foo:·\${bar.${baz}.bat}`);
  

```


